src/trajectory.jl

```diff
@@ -115,90 +117,152 @@ Multinomial sampler carried during the building of the tree.
 end
 
 """
 Slice sampler for the starting single leaf tree.
 Slice variable is initialized.
 """
-SliceTreeSampler(rng::AbstractRNG, H::AbstractFloat) = SliceTreeSampler(log(rand(rng)) - H, 1)
+SliceTreeSampler(rng::AbstractRNG, z0::PhasePoint) = SliceTreeSampler(z0, log(rand(rng)) - energy(z0), 1)
 
 """
 Multinomial sampler for the starting single leaf tree.
-Tree weight is just the probability of the only leave.
+Tree weight is the (unnormalised) energy of the leaf.
+
+Ref: https://github.com/stan-dev/stan/blob/develop/src/stan/mcmc/hmc/nuts/base_nuts.hpp#L226
 """
-MultinomialTreeSampler(rng::AbstractRNG, H::AbstractFloat) = MultinomialTreeSampler(-H)
+MultinomialTreeSampler(rng::AbstractRNG, z0::PhasePoint) = MultinomialTreeSampler(z0, zero(energy(z0)))
 
 """
 Create a slice sampler for a single leaf tree:
 - the slice variable is copied from the passed-in sampler `s` and
 - the number of acceptable candicates is computed by comparing the slice variable against the current energy.
 """
-makebase(s::SliceTreeSampler, H::AbstractFloat) = SliceTreeSampler(s.logu, (s.logu <= -H) ? 1 : 0)
+SliceTreeSampler(s::SliceTreeSampler, H0::AbstractFloat, zcand::PhasePoint) = 
+    SliceTreeSampler(zcand, s.logu, (s.logu <= -energy(zcand)) ? 1 : 0)
 
 """
-Create a multinomial sampler for a single leaf tree:
-- the tree weight is just the probability of the only leave.
+Multinomial sampler for the starting single leaf tree.
+- tree weight is the (unnormalised) energy of the leaf.
 """
-makebase(s::MultinomialTreeSampler, H::AbstractFloat) = MultinomialTreeSampler(-H)
+MultinomialTreeSampler(s::MultinomialTreeSampler, H0::AbstractFloat, zcand::PhasePoint) = 
+    MultinomialTreeSampler(zcand, H0 - energy(zcand))
+
+function combine(rng::AbstractRNG, s1::SliceTreeSampler, s2::SliceTreeSampler)
+    @assert s1.logu == s2.logu "Cannot combine two slice sampler with different slice variable"
+    n = s1.n + s2.n
+    zcand = rand(rng) < s1.n / n ? s1.zcand : s2.zcand
+    SliceTreeSampler(zcand, s1.logu, n)
+end
+
+function combine(zcand::PhasePoint, s1::SliceTreeSampler, s2::SliceTreeSampler)
+    @assert s1.logu == s2.logu "Cannot combine two slice sampler with different slice variable"
+    n = s1.n + s2.n
+    SliceTreeSampler(zcand, s1.logu, n)
+end
+
+function combine(rng::AbstractRNG, s1::MultinomialTreeSampler, s2::MultinomialTreeSampler)
+    logw = logaddexp(s1.logw, s2.logw)
+    zcand = rand(rng) < exp(s1.logw - logw) ? s1.zcand : s2.zcand
+    MultinomialTreeSampler(zcand, logw)
+end
+
+function combine(zcand::PhasePoint, s1::MultinomialTreeSampler, s2::MultinomialTreeSampler)
+    logw = logaddexp(s1.logw, s2.logw)
+    MultinomialTreeSampler(zcand, logw)
+end
+
+mh_accept(
+    rng::AbstractRNG,
+    s::SliceTreeSampler,
+    s′::SliceTreeSampler
+) = rand(rng) < min(1, s′.n / s.n)
+
+mh_accept(
+    rng::AbstractRNG,
+    s::MultinomialTreeSampler,
+    s′::MultinomialTreeSampler
+) = rand(rng) < exp(min(0, s′.logw - s.logw))
```

Comment:
Perhaps `exp(min(0, s′.logw - s.logw))` ==> `min(1, exp(s′.logw - s.logw))` for readbility.

Reply:
Could do this.